server-core/engine/src/main/java/io/infectnet/server/engine/content/system/creation/EntityCreatorSystem.java

```diff
@@ -0,0 +1,88 @@
+package io.infectnet.server.engine.content.system.creation;
+
+import io.infectnet.server.engine.core.entity.Category;
+import io.infectnet.server.engine.core.entity.Entity;
+import io.infectnet.server.engine.core.entity.EntityManager;
+import io.infectnet.server.engine.core.script.Request;
+import io.infectnet.server.engine.core.system.RequestOnlyProcessor;
+import io.infectnet.server.engine.core.util.ListenableQueue;
+import io.infectnet.server.engine.core.world.Position;
+import io.infectnet.server.engine.core.world.World;
+
+import java.util.Optional;
+
+
+public class EntityCreatorSystem extends RequestOnlyProcessor {
+
+  private final EntityManager entityManager;
+
+  private final World world;
+
+  public EntityCreatorSystem(EntityManager entityManager, World world) {
+    this.entityManager = entityManager;
+    this.world = world;
+  }
+
+  @Override
+  public void registerRequestListeners(ListenableQueue<Request> requestQueue) {
+    requestQueue.addListener(EntityCreationRequest.class, this::consumeEntityCreationRequest);
+  }
+
+  private void consumeEntityCreationRequest(Request request) {
+    EntityCreationRequest entityCreationRequest = (EntityCreationRequest) request;
+
+    // This Optional.get() is safe to do, because EntityCreationRequests always have targets.
+    // This is enforced by the constructor.
+    Entity creatorEntity = entityCreationRequest.getTarget().get();
+
+    Position createPosition = null;
+
+    if (creatorEntity.getTypeComponent().getCategory() == Category.BUILDING) {
+
+      Optional<Position> freePosition =
+          getFreePositionAround(creatorEntity.getPositionComponent().getPosition());
+
+      if (freePosition.isPresent()) {
+        createPosition = freePosition.get();
+      }
+
+    } else {
+      createPosition = creatorEntity.getPositionComponent().getPosition();
+    }
+
+    if (createPosition != null) {
```

Comment:
Can you place the act of  finding a suitable position and creating the actual entity into separate methods? That way this consumer would be much more readable.